website/content/en/blog/2025/dynamic-query-splitting.md

```diff
@@ -0,0 +1,181 @@
+
+![QuerySplittingAndVerticalSharding](/images/blog/2025/query-splitting-and-vertical-sharding.png)
+
+## Introducing Dynamic Query Splitting
+
+Cortex **dynamic query splitting** was introduced to address the limitations of static configurations. Instead of applying a fixed split interval uniformly across all queries, the dynamic logic computes an optimal split interval per query—as a multiple of the configured base interval—based on query semantics and configurable constraints. If **dynamic vertical sharding** is also enabled, then both split interval and vertical shard size will be dynamically adjusted for every query. 
+
+The goal is to maximize query parallelism through both horizontal splitting and vertical sharding, while staying within safe and configurable limits that prevent system overload or inefficiency. This is best explained by how dynamic splitting solves two problems:
+
+### 1. **Queuing and Merge Bottlenecks from Over-Splitting**
+
+While increasing parallelism through splitting and sharding is generally beneficial, it becomes counterproductive when the number of subqueries or shards far exceeds the number of available queriers.
+
+The number of splits when using a fixed interval increases with the query’s time range, which is under the user's control. For example:
+
+* With a static split interval of 24 hours, a 7-day query results in 7 horizontal splits
+* If the user increased query range to 100 days, the query is split into 100 horizontal splits.
+* If vertical sharding is also enabled and configured to use 5 shards (`query_vertical_shard_size: 5`), each split is further divided—leading to a total of 500 individual shards.
+
+When the number of shards grows too large:
+
+* Backend workers become saturated, causing subqueries to queue and increasing total latency.
+* The query-frontend merges hundreds of partial results, which introduces overhead that can outweigh the benefits of parallelism.
+* Overall system throughput degrades, especially when multiple large queries are executed concurrently.
+
+To solve this issue, a new configuration `max_shards_per_query` is introduced for the maximum parallelism per query. Given the same example above:
+
+* With a static split interval of 24 hours and `max_shards_per_query` set to 75, a 7-day query still results in 7 splits.
+* If the user increased query range to 100 days, the dynamic splitting algorithm will adjust the split interval to be 48 hours, producing 50 horizontal splits—keeping the total within the target of 75 shards.
+* If vertical sharding is enabled and configured to use 5 shards, the dynamic logic selects the optimal combination of split interval and vertical shards to maximize parallelism without exceeding 75 shards.
```

Comment:
Should we briefly explain how we calculate the optimal combination, or is that too much detail for this doc?